From 5b6d2e38fe1c1c2b85f8bcadc17616025fb0ea01 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:19 -0400
Subject: [usb] ch341: switch to generic TIOCMIWAIT implementation

Message-id: <1410887463-20674-248-git-send-email-dzickus@redhat.com>
Patchwork-id: 95282
O-Subject: [RHEL7 PATCH 247/591] USB: ch341: switch to generic TIOCMIWAIT implementation
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 5e409a265d9327689eef15a8fff1ffcb7f8fc2e1
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:32 2014 +0100

    USB: ch341: switch to generic TIOCMIWAIT implementation

    Switch to the generic TIOCMIWAIT implementation which does not suffer
    from the races involved when using the deprecated sleep_on functions.

    Acked-by: Arnd Bergmann <arnd@arndb.de>
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/ch341.c b/drivers/usb/serial/ch341.c
index acc8865..ac21618 100644
--- a/drivers/usb/serial/ch341.c
+++ b/drivers/usb/serial/ch341.c
@@ -463,7 +463,14 @@ static void ch341_update_line_status(struct usb_serial_port *port,
  if (!delta)
   return;
 
+ if (delta & CH341_BIT_CTS)
+  port->icount.cts++;
+ if (delta & CH341_BIT_DSR)
+  port->icount.dsr++;
+ if (delta & CH341_BIT_RI)
+  port->icount.rng++;
  if (delta & CH341_BIT_DCD) {
+  port->icount.dcd++;
   tty = tty_port_tty_get(&port->port);
   if (tty) {
    usb_serial_handle_dcd_change(port, tty,
@@ -510,46 +517,6 @@ exit:
    __func__, status);
 }
 
-static int ch341_tiocmiwait(struct tty_struct *tty, unsigned long arg)
-{
- struct usb_serial_port *port = tty->driver_data;
- struct ch341_private *priv = usb_get_serial_port_data(port);
- unsigned long flags;
- u8 prevstatus;
- u8 status;
- u8 changed;
-
- spin_lock_irqsave(&priv->lock, flags);
- prevstatus = priv->line_status;
- spin_unlock_irqrestore(&priv->lock, flags);
-
- for (;;) {
-  interruptible_sleep_on(&port->port.delta_msr_wait);
-  /* see if a signal did it */
-  if (signal_pending(current))
-   return -ERESTARTSYS;
-
-  if (port->serial->disconnected)
-   return -EIO;
-
-  spin_lock_irqsave(&priv->lock, flags);
-  status = priv->line_status;
-  spin_unlock_irqrestore(&priv->lock, flags);
-
-  changed = prevstatus ^ status;
-
-  if (((arg & TIOCM_RNG) && (changed & CH341_BIT_RI)) ||
-      ((arg & TIOCM_DSR) && (changed & CH341_BIT_DSR)) ||
-      ((arg & TIOCM_CD)  && (changed & CH341_BIT_DCD)) ||
-      ((arg & TIOCM_CTS) && (changed & CH341_BIT_CTS))) {
-   return 0;
-  }
-  prevstatus = status;
- }
-
- return 0;
-}
-
 static int ch341_tiocmget(struct tty_struct *tty)
 {
  struct usb_serial_port *port = tty->driver_data;
@@ -603,7 +570,7 @@ static struct usb_serial_driver ch341_device = {
  .break_ctl         = ch341_break_ctl,
  .tiocmget          = ch341_tiocmget,
  .tiocmset          = ch341_tiocmset,
- .tiocmiwait        = ch341_tiocmiwait,
+ .tiocmiwait        = usb_serial_generic_tiocmiwait,
  .read_int_callback = ch341_read_int_callback,
  .port_probe        = ch341_port_probe,
  .port_remove       = ch341_port_remove,
-- 
1.7.1